adapters/http-vertx/src/main/java/org/eclipse/hono/adapter/http/vertx/VertxBasedHttpProtocolAdapter.java

```diff
@@ -74,12 +93,17 @@ protected void addRoutes(final Router router) {
 
         if (getConfig().isAuthenticationRequired()) {
 
-            final HonoClientBasedAuthProvider authProvider = Optional.ofNullable(usernamePasswordAuthProvider)
-                .orElse(new UsernamePasswordAuthProvider(getCredentialsServiceClient(), getConfig()));
-
-            final Handler<RoutingContext> basicAuthHandler = new HonoBasicAuthHandler(authProvider, getConfig().getRealm());
-            addTelemetryApiRoutes(router, basicAuthHandler);
-            addEventApiRoutes(router, basicAuthHandler);
+            final ChainAuthHandler authHandler = new HonoChainAuthHandler();
+            authHandler.append(new X509AuthHandler(
+                    Optional.ofNullable(clientCertAuthProvider).orElse(
+                            new X509AuthProvider(getCredentialsServiceClient(), getConfig())),
+                    getTenantServiceClient()));
+            authHandler.append(new HonoBasicAuthHandler(
```

Comment:
I think it would make sense to allow disabling the "basic auth" provider here. My understanding is that the X509 handler it enough to authenticate a device. In the case you want to configure only X509, then it would still pass along requests the basic auth handler in case X509 certificates are invalid. Meaning that you could probe the credentials backend for username/password information.

Reply:
I think we can afford the luxury of waiting and seeing if the need arises. We can then still add a config property to selectively enable/disable auth mechanisms

Reply:
How would you define "need" here? Right now you allow access to the device registry, querying for username/password credentials although this protocol adapter instance may never want to accept those.

Reply:
This protocol adapter *by definition* allows devices to authenticate using a client certificate or username/password credentials. So, it's a feature, not a bug :-)

Reply:
I see … works-as-designed™ … then let's keep it that way for the moment :wink: